packages/compass-connections/src/stores/connections-store.spec.ts

```diff
@@ -75,6 +91,163 @@ describe('use-connections hook', function () {
+      );
+
+      expect(result.current.recentConnections).to.deep.equal([
+        { id: '5', lastUsed: new Date(1647020087551), connectionOptions },
+        { id: '2', lastUsed: new Date(1647020087550), connectionOptions },
+        { id: '4', connectionOptions },
+      ]);
+    });
+  });
+
+  describe('#connect', function () {
+    const maxRecents = 10;
+
+    it(`calls onConnected`, async function () {
+      const onConnected = sinon.spy();
+      const { result } = renderHook(() =>
+        useConnections({
+          onConnected,
+          connectionStorage: mockConnectionStorage,
+          connectFn: () => Promise.resolve({} as any),
+          appName: 'Test App Name',
+        })
+      );
+
+      await result.current.connect({
+        id: 'new',
+        connectionOptions: {
+          connectionString: 'mongodb://new-recent',
+        },
+      });
+
+      await waitFor(() => {
+        expect(onConnected).to.have.been.called;
+      });
+    });
+
+    it(`truncates recents to ${maxRecents}`, async function () {
+      const mockRecents = createMockRecents(maxRecents);
+
+      mockConnectionStorage.loadAll = () => Promise.resolve(mockRecents);
+
+      const { result } = renderHook(() =>
+        useConnections({
+          onConnected: noop,
+          connectionStorage: mockConnectionStorage,
+          connectFn: () => Promise.resolve({} as any),
+          appName: 'Test App Name',
+        })
+      );
+
+      // Wait for the async loading of connections to complete.
+      await waitFor(() =>
+        expect(result.current.state.connections.length).to.equal(
+          mockRecents.length
+        )
+      );
+
+      await result.current.connect({
+        id: 'new',
+        connectionOptions: {
+          connectionString: 'mongodb://new-recent',
+        },
+      });
+
+      await waitFor(() => {
+        // eslint-disable-next-line @typescript-eslint/unbound-method
+        expect(mockConnectionStorage.delete).to.have.been.calledWith(
+          mockRecents[mockRecents.length - 1]
+        );
+      });
+    });
+
+    it(`does not remove recent if there are less then ${maxRecents}`, async function () {
+      const mockRecents = createMockRecents(maxRecents - 1);
+
+      mockConnectionStorage.loadAll = () => Promise.resolve(mockRecents);
+
+      const onConnected = sinon.spy();
+      const { result } = renderHook(() =>
+        useConnections({
+          onConnected,
+          connectionStorage: mockConnectionStorage,
+          connectFn: () => Promise.resolve({} as any),
+          appName: 'Test App Name',
+        })
+      );
+
+      await result.current.connect({
+        id: 'new',
+        connectionOptions: {
+          connectionString: 'mongodb://new-recent',
+        },
+      });
+
+      // this may cause a false negative, but there is no other reliable way to
+      // test this case. It would fail eventually if the functionality is broken.
+      await util.promisify(setTimeout)(100);
```

Comment:
anything better?

Reply:
You can write the test in `onConnected` and use `done` similar to how we usually test callback behavior, right?

Reply:
If I understand it right, `onConnected` is called before any of this happens

Reply:
Would it make sense to move the callback and call it only after everything else finished there successfuly then maybe? Otherwise you're communicating that connection happened and then do a bunch of stuff that can fail in the same hook, is it sorta expected?

Reply:
Yeah we could do something Iike that. I think is intentional that one doesn't wait for the other though.

Probably better to have a `Promise.all()` to make it clear, but yeah if saving "recent" won't work users should still be able to connect and move on to the next screen.

@Anemy what's your feedback on this?

To be clear that test won't flake and it would quite likely fail all the time if a recent gets removed, but is not technically guaranteed to be red if the feature is broken.

Reply:
I see what you mean it's a bit hard to test.
I think it makes sense **not** to wait for a file system access or entry before moving the user past the connect screen? Probably super fast on most machines but could cause some delay.
Can't think of a cleaner way without bringing in extra logic or debugs in the flow, I'm alright with the setTimeout here as it's only in one test and will probably never fail, can always improve it later if somehow it does flake.

Reply:
> this flow has completed and then monitoring debug logs in the test work

Seems equally convoluted, If we have to change the code for this test maybe we could add an optional callback like `onConnectionProcessComplete` or something? That we could do, not really useful outside the tests but is kinda clean.